serialized_data_interface/local_sdi.py

```diff
@@ -19,42 +20,41 @@ def create_new_metadata(metadata):
     return metadata
 
 
-def change_zip_file(charm_name, metadata):
-    temp_dir = f"{charm_name}.tmp"
-    charm_zip = f"{charm_name}.charm"
+def change_zip_file(charm_name, metadata, charm_path="."):
+    charm_zip = f"{charm_path}/{charm_name}.charm"
 
-    os.mkdir(temp_dir)
+    with TemporaryDirectory() as temp_dir:
+        with ZipFileWithPermissions(charm_zip) as old_zip:
+            old_zip.extractall(path=temp_dir)
 
-    with ZipFileWithPermissions(f"{charm_name}.charm") as old_zip:
-        old_zip.extractall(path=temp_dir)
+        with ZipFile(charm_zip, "w", ZIP_DEFLATED) as new_zip:
+            for dirpath, dirnames, filenames in os.walk(temp_dir, followlinks=True):
+                dirpath = Path(dirpath)
+                for filename in filenames:
+                    filepath = dirpath / filename
+                    if "metadata.yaml" in filename:
+                        new_zip.writestr(
+                            str(filepath.relative_to(temp_dir)), yaml.dump(metadata)
+                        )
+                    else:
+                        new_zip.write(
+                            str(filepath), str(filepath.relative_to(temp_dir))
```

Comment:
It looks like `ZipFile.write` accepts `PosixPath`s here, so we can remove both calls to `str()`.